crates/networking/p2p/discv5/messages.rs

```diff
@@ -0,0 +1,1212 @@
+use std::{array::TryFromSliceError, net::IpAddr};
+
+use aes::cipher::{KeyIvInit, StreamCipher, StreamCipherError};
+use aes_gcm::{Aes128Gcm, KeyInit, aead::AeadMutInPlace};
+use bytes::{BufMut, Bytes};
+use ethrex_common::H256;
+use ethrex_rlp::{
+    decode::RLPDecode,
+    encode::RLPEncode,
+    error::RLPDecodeError,
+    structs::{Decoder, Encoder},
+};
+
+use crate::types::NodeRecord;
+
+type Aes128Ctr64BE = ctr::Ctr64BE<aes::Aes128>;
+
+// Max and min packet sizes as defined in
+// https://github.com/ethereum/devp2p/blob/master/discv5/discv5-wire.md#udp-communication
+// Used for package validation
+const MIN_PACKET_SIZE: usize = 63;
+const MAX_PACKET_SIZE: usize = 1280;
+/// 32 src-id + 1 sig-size + 1 eph-key-size
+const HANDSHAKE_AUTHDATA_HEAD: usize = 34;
+// protocol data
+const PROTOCOL_ID: &[u8] = b"discv5";
+const PROTOCOL_VERSION: u16 = 0x0001;
+// masking-iv size for a u128
+const IV_MASKING_SIZE: usize = 16;
+// static_header end limit: 23 bytes from static_header + 16 from iv_masking
+const STATIC_HEADER_END: usize = IV_MASKING_SIZE + 23;
+
+#[derive(Debug, thiserror::Error)]
+pub enum PacketDecodeErr {
+    #[error("RLP decoding error")]
+    RLPDecodeError(#[from] RLPDecodeError),
+    #[error("Invalid packet size")]
+    InvalidSize,
+    #[error("Invalid protocol: {0}")]
+    InvalidProtocol(String),
+    #[error("Stream Cipher Error: {0}")]
+    ChipherError(String),
+    #[error("TryFromSliceError: {0}")]
+    TryFromSliceError(#[from] TryFromSliceError),
+    #[error("Io Error: {0}")]
+    IoError(#[from] std::io::Error),
+}
+
+impl From<StreamCipherError> for PacketDecodeErr {
+    fn from(error: StreamCipherError) -> Self {
+        PacketDecodeErr::ChipherError(error.to_string())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Packet {
+    Ordinary(Ordinary),
+    WhoAreYou(WhoAreYou),
+    Handshake(Handshake),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PacketHeader {
+    pub static_header: Vec<u8>,
+    pub flag: u8,
+    pub nonce: Vec<u8>,
+    pub authdata: Vec<u8>,
+    /// Offset in the encoded packet where authdata ends, i.e where the header ends.
+    pub header_end_offset: usize,
+}
+
+impl Packet {
+    pub fn decode(
+        dest_id: &H256,
+        decrypt_key: &[u8],
```

Comment:
Are the keys actually variable size or could they be arrays/references to arrays?